TombEngine/Objects/TR5/Entity/tr5_imp.cpp

```diff
@@ -96,206 +135,252 @@ namespace TEN::Entities::Creatures::TR5
+				break;
 
-					if (AI.distance >= pow(SECTOR(0.5f), 2))
-					{
-						if (AI.distance < pow(SECTOR(2), 2))
-							item->Animation.TargetState = IMP_STATE_WALK;
-					}
-					else
-						item->Animation.TargetState = IMP_STATE_IDLE;
+			case IMP_STATE_IDLE:
+				creature->MaxTurn = 0;
+				creature->Flags = 0;
 
+				if (isScared)
+				{
+					item->Animation.TargetState = IMP_STATE_SCARED;
 					break;
+				}
 
-				case IMP_STATE_ATTACK_1:
-				case IMP_STATE_ATTACK_2:
-					creature->MaxTurn = -1;
+				if (AI.bite && AI.distance < IMP_IDLE_ATTACK_RANGE && item->TriggerFlags < 10)
+				{
+					if (GetRandomControl() & 1)
+						item->Animation.TargetState = IMP_STATE_ATTACK_1;
+					else
+						item->Animation.TargetState = IMP_STATE_JUMP_ATTACK;
+				}
+				else if (item->AIBits == FOLLOW)
+				{
+					item->Animation.TargetState = IMP_STATE_WALK;
+				}
+				else if(item->TriggerFlags == IMP_OCB_THROW_STONES)
+				{
+					item->Animation.TargetState = IMP_STATE_THROW_STONES;
+				}
+				else if (AI.distance > IMP_WALK_RANGE)
+				{
+					item->Animation.TargetState = IMP_STATE_RUN;
+				}
+				else if (AI.distance > IMP_LOW_ATTACK_RANGE || item->TriggerFlags < 10)
+				{
+					item->Animation.TargetState = IMP_STATE_WALK;
+				}
 
-					if (creature->Flags == 0 &&
-						item->TouchBits & 0x280)
-					{
-						DoDamage(creature->Enemy, 3);
-						CreatureEffect2(item, ImpBite, 10, item->Pose.Orientation.y, DoBloodSplat);
-					}
+				break;
 
-					break;
+			case IMP_STATE_RUN:
+				creature->MaxTurn = IMP_RUN_TURN_RATE_MAX;
 
-				case IMP_STATE_SCARED:
-					creature->MaxTurn = ANGLE(7.0f);
+				if (isScared)
+				{
+					item->Animation.TargetState = IMP_STATE_SCARED;
 					break;
+				}
 
-				case IMP_STATE_START_CLIMB:
-				case IMP_STATE_START_ROLL:
-					creature->MaxTurn = 0;
-					break;
+				if (AI.distance < IMP_LOW_ATTACK_RANGE)
+					item->Animation.TargetState = IMP_STATE_IDLE;
+				else if (AI.distance < IMP_WALK_RANGE)
+					item->Animation.TargetState = IMP_STATE_WALK;
 
-				case IMP_STATE_THROW_STONES:
-					creature->MaxTurn = -1;
+				break;
 
-					if (item->Animation.FrameNumber - g_Level.Anims[item->Animation.AnimNumber].frameBase == 40)
-						ImpThrowStones(item);
+			case IMP_STATE_ATTACK_1:
+				creature->MaxTurn = 0;
 
-					break;
+				if (!(creature->Flags & 1) &&
+					item->TouchBits.Test(ImpRightHandBite.meshNum))
+				{
+					DoDamage(creature->Enemy, IMP_ATTACK_DAMAGE);
+					CreatureEffect2(item, ImpRightHandBite, 10, item->Pose.Orientation.y, DoBloodSplat);
+					creature->Flags |= 1;
+				}
 
-				default:
-					break;
+				break;
+			case IMP_STATE_JUMP_ATTACK:
+				RotateTowardTarget(*item, AI, IMP_ATTACKJUMP_OR_STONE_TURN_RATE_MAX);
+
+				if (!(creature->Flags & 1) && item->TouchBits.Test(ImpRightHandBite.meshNum))
+				{
+					DoDamage(creature->Enemy, IMP_ATTACK_DAMAGE);
+					CreatureEffect(item, ImpRightHandBite, DoBloodSplat);
+					creature->Flags |= 1;
 				}
```

Comment:
"creature->Flags & 1" is just checking if the flags is "creature->Flags == 1"

and "creature->Flags |= 1" is the reseting the variable, the same as "creature->Flags = 0"

I know it could be done with a bool, but every entity (that do melee attacks) also use the Flags like an short, (someone even use "creature->Flags =2".

Not a big issue, I leave up to you if prefer changing it or leave it with bitwise operations.